### PR TITLE
Reorganize grammars

### PR DESCRIPTION
**Pull request by Feder1co5oave**
## Description

I noticed commonmark tests were being run in `gfm` mode. So I disabled it and moved block rules around to better represent the difference between the two grammar flavors:
- code fences are now parsed with `gfm: false` too (they're a part of commonmark)
- ATX headings require a space after the `#`s in `gfm: false` mode, too. They're still not required in `pedantic` mode.
- tables are supported whenever `gfm: true` (they're a part of gfm), so the old `tables` mode (described as "github flavored markdown + tables") is now practically the same as `gfm` mode.
- re-write the `paragraph` rule to make it more clear and efficient.

  - setext headings can no longer interrupt paragraphs. This means that the following markdown
    ```markdown
    Paragraph?
    Heading?
    =======
    ```
    was previously parsed (and still is, but only in `pedantic` mode) as 
    ```html
    <p>Paragraph?</p>
    <h1 id="heading">Heading?</h1>
    ```
    whereas now it is
    ```html
    <p>Paragraph?
    Heading?
    =======</p>
    ```
    Neither of these complies with commonmark, which introduced multi-line setext headings and whould parse it like
    ```html
    <h1>Paragraph?
    Heading?</h1>
    ```
    This involves parsing a paragraph-like block, followed by a line made of `=` or `-`. Now that I think of it, I'm not sure it's worth introducing this breaking change, _and then another_. Or we could merge this and wait until we get setext heading parsing 100% compliant before releasing. Let's hear opinions on this.

  - lists can now interrupt paragraphs, but only if unordered, or starting from 1. (as per [example 268](https://spec.commonmark.org/0.28/#example-268))

### Fixes
I don't know at this time of any filed issues solved by this PR.

## Expectation
marked should parse commonmark in `normal` mode (`gfm: false, tables: false`), and github flavored markdown in `gfm` mode, according to spec.
We should consider deprecating the `tables` option.

- [x] Test(s) exist to ensure functionality and minimize regression (if no tests added, list tests covering this PR); or,

## Committer

In most cases, this should be a different person than the contributor.

- [ ] Draft GitHub release notes have been updated.
- [ ] CI is green (no forced merge required).
- [ ] Merge PR

### CODE REVIEW

File: lib/marked.js [hunk 1]
```diff
@@ -69,11 +71,15 @@ block.html = edit(block.html, 'i')
...
-  .replace('heading', block.heading)
-  .replace('lheading', block.lheading)
-  .replace('tag', block._tag) // pars can be interrupted by type (6) html blocks
+  .replace('heading', ' {0,3}#{1,6} +')
+  .replace('|lheading', '') // setex headings don't interrupt commonmark paragraphs
+  .replace('blockquote', ' {0,3}>')
+  .replace('fences', ' {0,3}(?:`{3,}|~{3,})[^`\\n]*\\n')
+  .replace('list', ' {0,3}(?:[*+-]|1[.)]) ') // only lists starting from 1 can interrupt
+  .replace('html', '</?(?:tag)(?: +|\\n|/?>)|<(?:script|pre|style|!--)')
+    .replace('tag', block._tag) // pars can be interrupted by type (6) html blocks
```
[the remark this answers is not in the document]
Comment (styfle) [reply lib/marked.js]: Shouldn't the lint step catch this and fail CI?

[the remark this answers is not in the document]
Comment (Feder1co5oave) [reply lib/marked.js]: eslint is set to [warn on indentation](https://github.com/markedjs/marked/blob/88a856137646c183196ce96632c290d78ab27ee0/.eslintrc.json#L9). It still returns 0.

[the remark this answers is not in the document]
Comment (styfle) [reply lib/marked.js]: @UziTech I think that would be best because it would save time for reviewers like in this scenario.

**Comment by styfle** [comment]
@davisjam Can you review for redos vulnerabilities?

File: lib/marked.js [hunk 2]
```diff
@@ -14,9 +14,9 @@
 var block = {
   newline: /^\n+/,
   code: /^( {4}[^\n]+\n*)+/,
-  fences: noop,
+  fences: /^ {0,3}(`{3,}|~{3,})([^`\n]*)\n(?:|([\s\S]*?)\n)(?: {0,3}\1[~`]* *(?:\n+|$)|$)/,
```
Comment (Feder1co5oave) [inline lib/marked.js]: this rule was copied from block.gfm as-is.

**Comment by UziTech** [comment]
@Feder1co5oave could you fix the merge conflicts so we can get this merged?

**Pull request closed without merging**